File: training-data/c/prim_mst.c
#include <stdio.h>
#include <limits.h>

#define V 5

int min_key(int key[], int in_mst[]) {
    int min = INT_MAX, min_index = -1;
    for (int v = 0; v < V; v++) {
        if (!in_mst[v] && key[v] < min) {
            min = key[v];
            min_index = v;
        }
    }
    return min_index;
}

void prim_mst(int graph[V][V]) {
    int parent[V];
    int key[V];
    int in_mst[V] = {0};

    for (int i = 0; i < V; i++) key[i] = INT_MAX;
    key[0] = 0;
    parent[0] = -1;

    for (int count = 0; count < V - 1; count++) {
        int u = min_key(key, in_mst);
        in_mst[u] = 1;
        for (int v = 0; v < V; v++) {
            if (graph[u][v] && !in_mst[v] && graph[u][v] < key[v]) {
                key[v] = graph[u][v];
                parent[v] = u;
            }
        }
    }

    int total_weight = 0;
    for (int i = 1; i < V; i++) {
        printf("%d - %d : %d\n", parent[i], i, key[i]);
        total_weight += key[i];
    }
    printf("total: %d\n", total_weight);
}

int main(void) {
    int graph[V][V] = {
        {0, 2, 0, 6, 0},
        {2, 0, 3, 8, 5},
        {0, 3, 0, 0, 7},
        {6, 8, 0, 0, 9},
        {0, 5, 7, 9, 0}
    };
    prim_mst(graph);
    return 0;
}

File: training-data/fortran/prim_mst.f90
module prim_mst_mod
    implicit none
contains
    function prim_mst(graph, n) result(total_weight)
        integer, intent(in) :: graph(n, n)
        integer, intent(in) :: n
        integer :: total_weight
        integer, parameter :: inf = 999999
        integer :: key(n), parent(n)
        logical :: in_mst(n)
        integer :: i, j, u, min_key

        key = inf
        in_mst = .false.
        key(1) = 0
        parent(1) = 0
        total_weight = 0

        do i = 1, n
            min_key = inf
            u = -1
            do j = 1, n
                if (.not. in_mst(j) .and. key(j) < min_key) then
                    min_key = key(j)
                    u = j
                end if
            end do
            if (u == -1) exit
            in_mst(u) = .true.
            total_weight = total_weight + key(u)
            do j = 1, n
                if (graph(u, j) /= 0 .and. .not. in_mst(j) .and. graph(u, j) < key(j)) then
                    key(j) = graph(u, j)
                    parent(j) = u
                end if
            end do
        end do
    end function prim_mst
end module prim_mst_mod

program main
    use prim_mst_mod
    implicit none
    integer, parameter :: n = 5
    integer :: graph(n, n)

    graph = 0
    graph(1, 2) = 2; graph(2, 1) = 2
    graph(1, 3) = 3; graph(3, 1) = 3
    graph(2, 3) = 1; graph(3, 2) = 1
    graph(2, 4) = 4; graph(4, 2) = 4
    graph(3, 4) = 5; graph(4, 3) = 5
    graph(3, 5) = 6; graph(5, 3) = 6
    graph(4, 5) = 7; graph(5, 4) = 7

    print *, prim_mst(graph, n)
end program main
